crates/rust-analyzer/src/main_loop.rs

```diff
@@ -706,18 +717,10 @@ fn on_diagnostic_task(task: DiagnosticTask, msg_sender: &Sender<Message>, state:
     }
 }
 
-fn send_startup_progress(
-    sender: &Sender<Message>,
-    loop_state: &mut LoopState,
-    world_state: &WorldState,
-) {
-    if !world_state.feature_flags.get("notifications.workspace-loaded") {
-        return;
-    }
-
-    let total: usize = world_state.workspaces.iter().map(|it| it.n_packages()).sum();
+fn send_startup_progress(sender: &Sender<Message>, loop_state: &mut LoopState) {
+    let total: usize = loop_state.roots_total;
     let prev_progress = loop_state.roots_scanned_progress;
-    let progress = total - world_state.roots_to_scan;
+    let progress = total - loop_state.roots_to_scan;
     loop_state.roots_scanned_progress = Some(progress);
 
     match (prev_progress, loop_state.workspace_loaded) {
```

Comment:
It's not clear to me, but isn't `roots_to_scan` monotonically decreasing here? It feels like we shouldn't need the `progress != prev` check, nor the `None` value for the first call.

Reply:
It is monotonically decreasing / (or increasing if we changed to `roots_scanned`). But we still want to  keep check whether the progress is reported to client side and the first time of the progress reporting. So I think the `Option<usize>` and the `progress != prev` check is unavoidable ??

Reply:
I was thinking we could use `0` instead of `None`. If it's monotonic, the `progress != prev` test is always true?

Reply:
It is monotonic, but not strictly monotone such that that comparison will be equal occasionally. And note that the `prev`  is set like this :

```rust
let prev_progress = loop_state.roots_scanned_progress;
let progress = total - loop_state.roots_to_scan;
loop_state.roots_scanned_progress = Some(progress);
```

We changed roots_scanned_progress after we get the value so it will be different sometimes. (i.e. the ` loop_state.roots_to_scan` was changed in the loop)

Reply:
And one more thing, we can't guarantee `progress` will be zero when the first call of this function. So we have to use another variable to store that state. Anyway, I will rename it to `roots_progress_reported`.